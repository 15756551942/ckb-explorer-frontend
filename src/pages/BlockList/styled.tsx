import styled from 'styled-components'

export const BlockListPanel = styled.div`
  width: 100%;
  margin-top: 40px;
  margin-bottom: 40px;
  border-radius: 6px;
  overflow: hidden;
  box-shadow: 0 2px 6px 0 rgba(77, 77, 77, 0.21);

  @media (max-width: 700px) {
    margin-top: 0px;
    margin-bottom: 0px;
    padding: 0px 20px 0px 20px;

    .block__green__background {
      margin-left: -20px;
      height: 61px;
      width: calc(100% + 40px);
      background-color: #3cc68a;
      z-index: 1;
    }
  }
`

export const ContentTitle = styled.div`
  font-size: 50px;
  color: black;
  margin: 0 auto;
  text-align: center;

  @media (max-width: 700px) {
    font-size: 26px;
  }

  &:after {
    content: '';
    background: #3cc68a;
    height: 4px;
    width: 197px;
    display: block;
    margin: 0 auto;

    @media (max-width: 700px) {
      width: 80px;
    }
  }
`

export const ContentTable = styled.div`
  @media (max-width: 700px) {
    margin-top: -41px;
    z-index: 2;
  }
  overflow-x: auto;
`
export const CommonPagition = styled.div`
  .rc-pagination {
    height: 40px;
    display: flex;
    justify-content: center;
    align-items: center;
    li:focus {
      outline: none !important;
    }
    .rc-pagination-item-active {
      background-color: #4bbc8e !important;
      border-color: #4bbc8e !important;
    }
    .rc-pagination-item {
      &:hover {
        border-color: #4bbc8e !important;
      }
    }
    .rc-pagination-options {
      .rc-pagination-options-quick-jumper {
        input:hover {
          border-color: #4bbc8e !important;
        }
      }
    }
  }
`

export const BlocksPagition = styled(CommonPagition)`
  margin-top: 60px;
  @media (max-width: 700px) {
    margin-top: 15px;
    margin-left: 3%;
  }
  width: 100%;
  overflow-x: auto;
`
