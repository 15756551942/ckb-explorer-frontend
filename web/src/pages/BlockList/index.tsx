import React, { useState } from 'react'
import { RouteComponentProps, Link } from 'react-router-dom'
import Pagination from 'rc-pagination'
import 'rc-pagination/assets/index.css'
import {
  BlockListPanel,
  ContentTitle,
  ContentTable,
  TableTitleRow,
  TableContentRow,
  TableMinerContentPanel,
  BlocksPagition,
} from './index.css'
import { parseDate } from '../../utils/date'
import Page from '../../components/Page'
import Header from '../../components/Header'
import Content from '../../components/Content'
import Footer from '../../components/Footer'
import BlockHeightIcon from '../../asserts/block_height.png'
import TransactionIcon from '../../asserts/transactions.png'
import CellConsumedIcon from '../../asserts/cell_consumed.png'
import MinerIcon from '../../asserts/miner.png'
import TimestampIcon from '../../asserts/timestamp.png'
import BlocksData from './mock'

const TableTitleItem = ({ image, title }: { image: string; title: string }) => {
  return (
    <div>
      <div>
        <img src={image} alt={title} />
        <div>{title}</div>
      </div>
    </div>
  )
}

const TableContentItem = ({ color, content }: { color: string; content: string }) => {
  return (
    <div
      style={{
        color,
        height: 78,
      }}
    >
      {content}
    </div>
  )
}

const TableMinerContentItem = ({ color, content }: { color: string; content: string }) => {
  return (
    <TableMinerContentPanel color={color}>
      <Link className="table__miner__content" to={`/address/${content}`}>
        {content}
      </Link>
    </TableMinerContentPanel>
  )
}

export default (props: React.PropsWithoutRef<RouteComponentProps<{ pageNo: string; pageSize: string }>>) => {
  const [clickableColor, normalColor] = ['#4bbc8e', '888888']

  const { match } = props
  const { params } = match
  const { pageNo, pageSize } = params

  const [currentPageNo, setCurrentPageNo] = useState(pageNo === undefined ? 1 : parseInt(pageNo, 10))
  const [currentPageSize, setCurrentPageSize] = useState(pageSize === undefined ? 3 : parseInt(pageSize, 10))

  // TODO: fetch transaction data from server
  const getBlocks = (page: number, size: number) => {
    return BlocksData.data.slice((page - 1) * size, page * size)
  }

  const onChange = (page: number, size: number) => {
    setCurrentPageNo(page)
    setCurrentPageSize(size)
    getBlocks(page, size)
  }

  return (
    <Page>
      <Header />
      <Content>
        <BlockListPanel width={window.innerWidth}>
          <ContentTitle>Blocks</ContentTitle>

          <ContentTable>
            <div>
              <TableTitleRow>
                <TableTitleItem image={BlockHeightIcon} title="Height" />
                <TableTitleItem image={TransactionIcon} title="Transactions" />
                <TableTitleItem image={CellConsumedIcon} title="Cell Consumed(B)" />
                <TableTitleItem image={MinerIcon} title="Miner" />
                <TableTitleItem image={TimestampIcon} title="Time" />
              </TableTitleRow>
              {getBlocks(currentPageNo, currentPageSize).map((data: any) => {
                return (
                  <TableContentRow key={data.block_hash}>
                    <TableContentItem color={clickableColor} content={data.number} />
                    <TableContentItem color={normalColor} content={data.transactions_count} />
                    <TableContentItem color={normalColor} content={data.cell_consumed} />
                    <TableMinerContentItem color={clickableColor} content={data.miner_hash} />
                    <TableContentItem color={normalColor} content={parseDate(data.timestamp)} />
                  </TableContentRow>
                )
              })}
            </div>
          </ContentTable>
          <BlocksPagition>
            <Pagination
              showQuickJumper
              showSizeChanger
              defaultPageSize={currentPageSize}
              defaultCurrent={currentPageNo}
              total={BlocksData.data.length}
              onChange={onChange}
            />
          </BlocksPagition>
        </BlockListPanel>
      </Content>
      <Footer />
    </Page>
  )
}
