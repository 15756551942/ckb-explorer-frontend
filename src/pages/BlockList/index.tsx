import React, { useEffect, useContext, Fragment, useMemo } from 'react'
import { RouteComponentProps, Link } from 'react-router-dom'
import queryString from 'query-string'
import { useTranslation } from 'react-i18next'
import { parseSimpleDate } from '../../utils/date'
import { BlockListPanel, ContentTable, HighLightValue } from './styled'
import Content from '../../components/Content'
import {
  TableTitleRow,
  TableTitleItem,
  TableContentRow,
  TableContentItem,
  TableMinerContentItem,
} from '../../components/Table'
import { shannonToCkb } from '../../utils/util'
import { parsePageNumber, adaptMobileEllipsis } from '../../utils/string'
import { BlockListPageParams, DELAY_BLOCK_NUMBER } from '../../utils/const'
import { localeNumberString } from '../../utils/number'
import { isMobile } from '../../utils/screen'
import { StateWithDispatch } from '../../contexts/providers/reducer'
import i18n from '../../utils/i18n'
import Pagination from '../../components/Pagination'
import OverviewCard, { OverviewItemData } from '../../components/Card/OverviewCard'
import { AppContext } from '../../contexts/providers'
import { getBlocks } from '../../service/app/block'

const BlockValueItem = ({ value, to }: { value: string; to: string }) => {
  return (
    <HighLightValue>
      <Link to={to}>
        <span>{value}</span>
      </Link>
    </HighLightValue>
  )
}

interface TableTitleData {
  title: string
  width: string
}

interface TableContentData {
  width: string
  to?: any
  content: string
}

const getTableContentDataList = (block: State.Block, index: number) => {
  const blockReward = `${localeNumberString(shannonToCkb(block.reward))}${index <= DELAY_BLOCK_NUMBER ? ' +' : ''}`
  return [
    {
      width: '14%',
      to: `/block/${block.number}`,
      content: localeNumberString(block.number),
    },
    {
      width: '14%',
      content: `${block.transactionsCount}`,
    },
    {
      width: '20%',
      content: blockReward,
    },
    {
      width: '37%',
      content: block.minerHash,
    },
    {
      width: '15%',
      content: parseSimpleDate(block.timestamp),
    },
  ] as TableContentData[]
}

const BlockCardItems = (block: State.Block, index: number) => {
  const blockReward = `${localeNumberString(shannonToCkb(block.reward))}${index <= DELAY_BLOCK_NUMBER ? ' +' : ''}`
  return [
    {
      title: i18n.t('home.height'),
      content: <BlockValueItem value={localeNumberString(block.number)} to={`/block/${block.number}`} />,
    },
    {
      title: i18n.t('home.transactions'),
      content: localeNumberString(block.transactionsCount),
    },
    {
      title: i18n.t('home.block_reward'),
      content: blockReward,
    },
    {
      title: i18n.t('block.miner'),
      content: <BlockValueItem value={adaptMobileEllipsis(block.minerHash, 12)} to={`/address/${block.minerHash}`} />,
    },
    {
      title: i18n.t('home.time'),
      content: parseSimpleDate(block.timestamp),
    },
  ] as OverviewItemData[]
}

export default ({
  dispatch,
  history: { replace, push },
  location: { search },
}: React.PropsWithoutRef<StateWithDispatch & RouteComponentProps>) => {
  const [t] = useTranslation()
  const TableTitles = useMemo(() => {
    return [
      {
        title: t('home.height'),
        width: '14%',
      },
      {
        title: t('home.transactions'),
        width: '14%',
      },
      {
        title: t('home.block_reward'),
        width: '20%',
      },
      {
        title: t('block.miner'),
        width: '37%',
      },
      {
        title: t('home.time'),
        width: '15%',
      },
    ]
  }, [t])

  const parsed = queryString.parse(search)
  const { blockListState } = useContext(AppContext)
  const { blocks = [] } = blockListState

  const currentPage = parsePageNumber(parsed.page, BlockListPageParams.PageNo)
  const pageSize = parsePageNumber(parsed.size, BlockListPageParams.PageSize)
  const totalPages = Math.ceil(blockListState.total / pageSize)

  useEffect(() => {
    if (pageSize > BlockListPageParams.MaxPageSize) {
      replace(`/block/list?page=${currentPage}&size=${BlockListPageParams.MaxPageSize}`)
    }
    getBlocks(currentPage, pageSize, dispatch)
  }, [replace, currentPage, pageSize, dispatch])

  const onChange = (page: number) => {
    push(`/block/list?page=${page}&size=${pageSize}`)
  }

  return (
    <Content>
      <BlockListPanel className="container">
        <div className="block__green__background" />
        {isMobile() ? (
          <ContentTable>
            <div className="block__panel">
              {blocks.map((block: State.Block, index: number) => {
                return <OverviewCard key={block.number} items={BlockCardItems(block, index)} />
              })}
            </div>
          </ContentTable>
        ) : (
          <ContentTable>
            <TableTitleRow>
              {TableTitles.map((data: TableTitleData) => {
                return <TableTitleItem width={data.width} title={data.title} key={data.title} />
              })}
            </TableTitleRow>
            {blocks.map((block: State.Block, blockIndex: number) => {
              return (
                block && (
                  <TableContentRow key={block.number} onClick={() => push(`/block/${block.blockHash}`)}>
                    {getTableContentDataList(block, blockIndex).map((data: TableContentData, index: number) => {
                      const key = index
                      return (
                        <Fragment key={key}>
                          {data.content === block.minerHash ? (
                            <TableMinerContentItem width={data.width} content={data.content} />
                          ) : (
                            <TableContentItem width={data.width} content={data.content} to={data.to} />
                          )}
                        </Fragment>
                      )
                    })}
                  </TableContentRow>
                )
              )
            })}
          </ContentTable>
        )}
        <div className="block_list__pagination">
          <Pagination currentPage={currentPage} totalPages={totalPages} onChange={onChange} />
        </div>
      </BlockListPanel>
    </Content>
  )
}
