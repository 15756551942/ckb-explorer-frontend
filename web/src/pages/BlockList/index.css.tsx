import styled from 'styled-components'

export const BlockListPanel = styled.div`
  width: 100%;
  margin-top: ${(props: { width: number }) => (150 * props.width) / 1920}px;
  margin-bottom: ${(props: { width: number }) => (200 * props.width) / 1920}px;
`

export const ContentTitle = styled.div`
  font-size: 50px;
  color: black;
  margin: 0 auto;
  text-align: center;

  &:after {
    content: '';
    background: #46ab81;
    height: 4px;
    width: 197px;
    display: block;
    margin: 0 auto;
  }
`

export const ContentTable = styled.div`
  margin-top: 58px;
  overflow-x: auto;
`

export const BlocksPagition = styled.div`
  margin-top: 60px;
  display: flex;
  align-items: center;
  justify-content: center;
`
