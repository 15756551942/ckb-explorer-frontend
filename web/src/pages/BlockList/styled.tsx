import styled from 'styled-components'

export const BlockListPanel = styled.div`
  width: 100%;
  margin-top: ${(props: { width: number }) => (150 * props.width) / 1920}px;
  margin-bottom: ${(props: { width: number }) => (200 * props.width) / 1920}px;
`

export const ContentTitle = styled.div`
  font-size: 50px;
  color: black;
  margin: 0 auto;
  text-align: center;

  &:after {
    content: '';
    background: #3CC68A;
    height: 4px;
    width: 197px;
    display: block;
    margin: 0 auto;
  }
`

export const ContentTable = styled.div`
  margin-top: 58px;
  overflow-x: auto;
`
export const CommonPagition = styled.div`
  .rc-pagination {
    min-width: 450px;
    height: 40px;
    display: flex;
    justify-content: center;
    align-items: center;
    li:focus {
      outline: none !important;
    }
    .rc-pagination-item-active {
      background-color: #4bbc8e !important;
    }
    .rc-pagination-item {
      &:hover {
        border-color: #4bbc8e !important;
      }
    }
    .rc-pagination-options {
      .rc-pagination-options-quick-jumper {
        input:hover {
          border-color: #4bbc8e !important;
        }
      }
    }
  }
`

export const BlocksPagition = styled(CommonPagition)`
  margin-top: 60px;
  width: 100%;
  overflow-x: auto;
`
