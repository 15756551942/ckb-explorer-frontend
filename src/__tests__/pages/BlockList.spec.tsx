import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import BlockList from '../../pages/BlockList'

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    length: 13,
    push: jest.fn(),
    block: jest.fn(),
    createHref: jest.fn(),
    go: jest.fn(),
    goBack: jest.fn(),
    goForward: jest.fn(),
    liten: jest.fn(),
    replace: jest.fn(),
    action: 'REPLACE',
    location: null,
  }),
  useLocation: () => ({
    pathname: '/welcome',
    hash: '',
    search: '',
    state: '',
  }),
}))

describe('Active Page', () => {
  it('is defined', () => {
    const app = shallow(<BlockList />)

    expect(app).toBeDefined()
  })

  it('shallow renders with empty data', async () => {
    const wrapper = renderer.create(<BlockList />).toJSON()
    expect(wrapper).toMatchSnapshot()
  })
})
