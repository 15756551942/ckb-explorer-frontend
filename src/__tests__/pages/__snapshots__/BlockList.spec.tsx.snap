// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active Page shallow renders with empty data 1`] = `
<div
  className="sc-fznJRM LSUZK"
>
  <div
    className="sc-AxjAm bAbQHZ container"
  >
    <div
      className="block__green__background"
    />
    <div
      className="sc-AxiKw esumCH"
    >
      <div
        className="sc-fznWqX hHLogm"
      >
        <div
          className="sc-fzoiQi fTOqfe"
          width="14%"
        >
          <div>
            Height
          </div>
        </div>
        <div
          className="sc-fzoiQi bfJcav"
          width="8%"
        >
          <div>
            Transactions
          </div>
        </div>
        <div
          className="sc-fzoiQi cEyfUN"
          width="20%"
        >
          <div>
            Reward (CKB)
          </div>
        </div>
        <div
          className="sc-fzoiQi igbORU"
          width="43%"
        >
          <div>
            Miner
          </div>
        </div>
        <div
          className="sc-fzoiQi fdZMGj"
          width="15%"
        >
          <div>
            Time
          </div>
        </div>
      </div>
    </div>
    <div
      className="block_list__pagination"
    >
      <div
        className="sc-fzoXWK ktWbCE"
      >
        <div
          className="sc-fzpmMD OKRCg"
        >
          <button
            onClick={[Function]}
          >
            First
          </button>
          <button
            className="pagination__left__button"
            onClick={[Function]}
            type="button"
          />
          <div
            className="pagination__middle__label"
          >
            Page 0 of 1 
          </div>
          <button
            className="pagination__right__button"
            onClick={[Function]}
            type="button"
          />
          <button
            className="pagination__last"
            onClick={[Function]}
            type="button"
          >
            Last
          </button>
        </div>
        <div
          className="sc-fznxKY kvnvQt"
        >
          <div
            className="pagination__page"
          >
            Page
          </div>
          <input
            className="pagination__input__page"
            onChange={[Function]}
            onKeyUp={[Function]}
            pattern="[0-9]*"
            type="text"
            value={2}
          />
          <button
            className="pagination__goto__page"
            onClick={[Function]}
            type="button"
          >
            Goto
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
